asdf/schema.py

```diff
@@ -547,30 +550,59 @@ def get_validator(schema={}, ctx=None, validators=None, url_mapping=None,
     return validator
 
 
-def validate_large_literals(instance, reading=False):
+def _validate_large_literals(instance, reading):
     """
     Validate that the tree has no large numeric literals.
     """
-    # We can count on 52 bits of precision
-    for instance in treeutil.iter_tree(instance):
+    def _validate(value):
+        # We can count on 52 bits of precision
+        if value <= ((1 << 51) - 1) and value >= -((1 << 51) - 2):
+            return
+
+        if reading:
+            warnings.warn(
+                f"Invalid integer literal value {value} detected while reading file. "
+                "The value has been read safely, but the file should be "
+                "fixed.",
+                AsdfWarning
+            )
+        else:
+            raise ValidationError(
+                f"Integer value {value} is too large to safely represent as a "
+                "literal in ASDF"
+            )
 
-        if not isinstance(instance, Integral):
-            continue
+    if isinstance(instance, Integral):
+        _validate(instance)
+    elif isinstance(instance, Mapping):
+        for key in instance:
+            if isinstance(key, Integral):
+                _validate(key)
 
-        if instance <= ((1 << 51) - 1) and instance >= -((1 << 51) - 2):
-            continue
 
-        if not reading:
-            raise ValidationError(
-                "Integer value {0} is too large to safely represent as a "
-                "literal in ASDF".format(instance))
+def _validate_mapping_keys(instance, reading):
+    """
+    Validate that mappings do not contain illegal key types
+    (as of ASDF Standard 1.6.0, only str, int, and bool are
+    permitted).
+    """
+    if not isinstance(instance, Mapping):
+        return
 
-        warnings.warn(
-            "Invalid integer literal value {0} detected while reading file. "
-            "The value has been read safely, but the file should be "
-            "fixed.".format(instance),
-            AsdfWarning
-        )
+    for key in instance:
+        if isinstance(key, tagged.Tagged) or not isinstance(key, (str, int, bool)):
+            if reading:
```

Comment:
If it accepts the bad key, what ends up in the python representation?  If it is a key that is not acceptable in Python, will that just trigger a different exception?

Reply:
If it's a float or a tagged string, then the value will end up in the tree.  If it's something unhashable, then it'll fail when pyyaml attempts to construct the tagged dict.  Maybe we should just fail on read as well?  I copied the pattern we had for large integers, but maybe it doesn't make sense to accept the key with a warning.

Reply:
There is the warning, but on the other hand I worry that many users won't see that in front of the traceback and know that is the root cause. Some options include more explanation that this is likely to cause an exception anyway (meh), or try to hash the key, and if not just throw the exception there. (is the second easy?)

Reply:
Unfortunately the pyyaml failure occurs before this method is even called, so the user won't get a chance to see the warning in the case of an unhashable key.  If the user does get to the warning then I think there isn't likely to be an exception.

Reply:
Well, in that case, there isn't much point in changing it. I wonder if there is a way to intercept it earlier, but it isn't a high priority. Should an issue be created? (no if it is near impossible).

Reply:
We already have our claws sunk firmly into pyyaml via our custom loader, so it wouldn't be much extra effort to catch this case in there.  I'll create an issue.

Reply:
https://github.com/asdf-format/asdf/issues/868